src/aten/Activation.cpp

```diff
@@ -158,4 +160,90 @@ Tensor& XPUNativeFunctions::gelu_backward_out(
   return grad_input;
 }
 
+Tensor XPUNativeFunctions::hardtanh(
+    const Tensor& self,
+    const Scalar& min,
+    const Scalar& max) {
+  Tensor result = at::empty_like(self);
+  return at::hardtanh_out(result, self, min, max);
+}
+
+Tensor& XPUNativeFunctions::hardtanh_out(
+    const Tensor& self,
+    const Scalar& min,
+    const Scalar& max,
+    Tensor& result) {
+  TORCH_CHECK(
+      self.scalar_type() != at::kBool, "Boolean inputs not supported for hardtanh");
+  Scalar min_, max_;
+  if (at::isIntegralType(self.scalar_type(), /*include_bool*/false)) {
+    int64_t minval = min.toLong();
+    int64_t maxval = max.toLong();
+    TORCH_CHECK(
+      self.dtype() != at::kByte || (minval >= 0 && maxval >=0), 
+      "cannot do hardtanh on an unsigned type with negative limits");
+    min_ = minval;
+    max_ = maxval;
+  } else {
+    min_ = min;
+    max_ = max;
+  }
+  return at::clamp_out(result, self, min_, max_);
+}
+
+Tensor& XPUNativeFunctions::hardtanh_(
+    Tensor& self,
+    const Scalar& min,
+    const Scalar& max) {
+  return at::hardtanh_out(self, self, min, max);
+}
+
+Tensor& XPUNativeFunctions::hardtanh_backward_out(
+    const Tensor& grad_output,
+    const Tensor& self,
+    const Scalar& min,
+    const Scalar& max,
+    Tensor& grad_input) {
+  auto iter = TensorIterator::borrowing_binary_op(grad_input, grad_output, self);
```

Comment:
Please also review CUDA generated code to rebase all operators level codes. `build/aten/src/ATen/RegisterCUDA.cpp`. The file is generated at building time. You can build a PyTorch with CUDA or review a built project.
<img width="1234" alt="image" src="https://github.com/intel/torch-xpu-ops/assets/28250760/d21bab34-8732-416b-bedc-5f133aed1968">